qgis, qgis-ltr: fix dependencies on darwin

willcohen opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Progress toward #71398. Builds on #150286, #150595, #156747, #156809, #156828.

###### Things done

@NixOS/darwin-maintainers I'm struggling with getting the final build of this to link and load correctly, though all the dependencies should finally be fixed. Even after fully disabling all of the `makeWrapper` work on darwin, I'm still getting mired down by being unable to get the dylibs to load. Pasting the end result here since building this derivation takes a while. 

Any suggestions would be MUCH appreciated.

```
$ /nix/store/fcr2r0dihgkqid9jv2c226k8jcpfvw1q-qgis-unwrapped-3.22.3/QGIS.app/Contents/MacOS/QGIS  
dyld[39783]: Library not loaded: @executable_path/lib/libqgis_app.3.22.3.dylib
  Referenced from: /nix/store/fcr2r0dihgkqid9jv2c226k8jcpfvw1q-qgis-unwrapped-3.22.3/QGIS.app/Contents/MacOS/.QGIS-wrapped
  Reason: tried: '/nix/store/fcr2r0dihgkqid9jv2c226k8jcpfvw1q-qgis-unwrapped-3.22.3/QGIS.app/Contents/MacOS/lib/libqgis_app.3.22.3.dylib' (not a mach-o file), '/usr/local/lib/libqgis_app.3.22.3.dylib' (no such file), '/usr/lib/libqgis_app.3.22.3.dylib' (no such file)
[1]    39783 abort      
➜  lib ls -al /nix/store/fcr2r0dihgkqid9jv2c226k8jcpfvw1q-qgis-unwrapped-3.22.3/QGIS.app/Contents/MacOS/lib 
total 26624
dr-xr-xr-x  9 root  wheel       288 Dec 31  1969 .
dr-xr-xr-x  8 root  wheel       256 Dec 31  1969 ..
-r-xr-xr-x  1 root  wheel  13502764 Dec 31  1969 .libqgis_app.3.22.3.dylib-wrapped
-r-xr-xr-x  1 root  wheel    115896 Dec 31  1969 .libqgispython.3.22.3.dylib-wrapped
-r-xr-xr-x  1 root  wheel      2485 Dec 31  1969 libqgis_app.3.22.3.dylib
lrwxr-xr-x  1 root  wheel        24 Dec 31  1969 libqgis_app.dylib -> libqgis_app.3.22.3.dylib
-r-xr-xr-x  1 root  wheel      2487 Dec 31  1969 libqgispython.3.22.3.dylib
lrwxr-xr-x  1 root  wheel        26 Dec 31  1969 libqgispython.dylib -> libqgispython.3.22.3.dylib
dr-xr-xr-x  5 root  wheel       160 Dec 31  1969 qgis
```

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

hexagonal-sun commented on pkgs/applications/gis/qgis/unwrapped.nix:
```diff
@@ -130,24 +137,31 @@ in mkDerivation rec {
     homepage = "https://www.qgis.org";
     license = lib.licenses.gpl2Plus;
-    platforms = with lib.platforms; linux;
+    platforms = with lib.platforms; linux ++ darwin;
```
I'd make this a lteral: `[ liunx darwin ]`.

hexagonal-sun commented on pkgs/applications/gis/qgis/unwrapped-ltr.nix:
```diff
@@ -131,18 +138,24 @@ in mkDerivation rec {
       --prefix PATH : ${lib.makeBinPath [ grass ]}
+  '' + lib.optionalString stdenv.isDarwin ''
+    wrapProgram $out/QGIS.app/Contents/MacOS/QGIS \
+      --prefix PATH : ${lib.makeBinPath [ grass ]}
```
Perhaps we can avoid the duplication here by setting a `exe` variable, which is platform dependent, the referencing that in `wrapProgram`?

willcohen replied:
On this point, even without being able to build, @hexagonal-sun do you know if there is a way to disable the automatic wrapping of all the dylibs without removing `makeWrapper` entirely? (I assume that you may have dealt with this in #131891). If not, to actually get this to work I may actually need to remove makeWrapper and just manually do what make-wrapper.sh does, so it only applies to this one executable, rather than to all the files in the .app.

hexagonal-sun commented on pkgs/applications/gis/qgis/unwrapped-ltr.nix:
```diff
@@ -131,18 +138,24 @@ in mkDerivation rec {
     homepage = "https://www.qgis.org";
     license = licenses.gpl2Plus;
-    platforms = platforms.linux;
+    platforms = platforms.linux ++ platforms.darwin;
```
I'd use: `with platforms; [ linux darwin ];`

hexagonal-sun reviewed (commented):
Some minor code cleanup.  Unfortunetly, I'm running an M1 and we don't have Qt support yet so I can't build on this end.

siraben commented:
Result of `nixpkgs-review pr 157862` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>qgis</li>
    <li>qgis-ltr</li>
  </ul>
</details>

siraben commented:
Result of `nixpkgs-review pr 157862` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>qgis</li>
    <li>qgis-ltr</li>
  </ul>
</details>

willcohen commented:
This does not fully complete the build for darwin, but does fix all of the dependencies. A .app does get produced at this point, but the dylibs still link incorrectly. My preference would be to merge this reduced-scope PR first a progress step since it deals with the underlying build requirements, and then I can still try to work out fixing the .app bundle's internal issues in a separate PR without having to pull these changes along with, because I don't expect them to change any more than as is.

willcohen commented:
As context, even the [official build recipe](https://github.com/qgis/QGIS-Mac-Packager/blob/master/qgis_bundle/qgis_bundle.bash) for the Mac .app has a whole bunch of steps needed to get the .app bundle working, so some substantial variant of these kinds of steps will probably eventually be needed for the nixpkgs bundle.

willcohen commented:
Force-pushed to remove darwin as a supported platform (since the .app isn't yet functional), but otherwise the same.

ttuegel commented:
> Would you suggest that the best course of action here is to just set dontWrapQtApps on darwin and manually run wrapQtApp as needed?

I would suggest setting `dontWrapQtApps` on Darwin. That script will only work correctly for ELF objects on Linux.

willcohen commented:
With #209801 merged it's a chance to dust this off. Rebased to master and fixed the dependencies and linking issues from before. However, attempting to load the QGIS.app for either qgis or qgis-ltr fails with the following crash report. Any suggestions @NixOS/darwin-maintainers?

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               QGIS [45199]
Path:                  /Volumes/VOLUME/*/QGIS.app/Contents/MacOS/QGIS
Identifier:            org.qgis.qgis3
Version:               3.28.2 (3.28.2 (exported))
Code Type:             ARM-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2023-01-11 11:18:43.4045 -0500
OS Version:            macOS 13.1 (22C65)
Report Version:        12
Anonymous UUID:        EFA3F67E-B6A3-CBFF-B0CE-8139B7515F3A

Sleep/Wake UUID:       C5799FA7-6E4C-4286-B156-47BAA862AF96

Time Awake Since Boot: 1300000 seconds
Time Since Wake:       47222 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Application Specific Information:
abort() called


Kernel Triage:
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x19fa4e1b0 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x19fa84cec pthread_kill + 288
2   libsystem_c.dylib             	       0x19f9be2c8 abort + 180
3   ???                           	       0x1020b0810 myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&) + 1052
4   ???                           	       0x1043a0948 ???
5   ???                           	       0x10439ba30 ???
6   ???                           	       0x104664064 QMessageLogger::fatal(char const*, ...) const + 56
7   ???                           	       0x103e6ef5c QGuiApplicationPrivate::createPlatformIntegration() + 7296
8   ???                           	       0x103e6ef80 QGuiApplicationPrivate::createEventDispatcher() + 36
9   ???                           	       0x10456c554 QCoreApplicationPrivate::init() + 1472
10  ???                           	       0x103e692ec QGuiApplicationPrivate::init() + 68
11  ???                           	       0x102a2c5fc QApplicationPrivate::init() + 52
12  ???                           	       0x102a2c5b0 QApplication::QApplication(int&, char**, int) + 164
13  ???                           	       0x10a22d0c4 QgsApplication::QgsApplication(int&, char**, bool, QString const&, QString const&) + 60
14  ???                           	       0x1020b335c main + 11012
15  dyld                          	       0x19f75be50 start + 2544

Thread 1:
0   libsystem_pthread.dylib       	       0x19fa7fe18 start_wqthread + 0

Thread 2:
0   libsystem_pthread.dylib       	       0x19fa7fe18 start_wqthread + 0

Thread 3:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 4:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 5:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 6:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 7:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 8:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 9:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x1167edb34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 10:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 11:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 12:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 13:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 14:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 15:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8

Thread 16:
0   libsystem_kernel.dylib        	       0x19fa49564 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x19fa85638 _pthread_cond_wait + 1232
2   ???                           	       0x117da1b34 blas_thread_server + 360
3   libsystem_pthread.dylib       	       0x19fa8506c _pthread_start + 148
4   libsystem_pthread.dylib       	       0x19fa7fe2c thread_start + 8


Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000000000   x1: 0x0000000000000000   x2: 0x0000000000000000   x3: 0x0000000000000000
    x4: 0x0000000010000800   x5: 0x000000000000001c   x6: 0x0000000000000000   x7: 0x0000000000011000
    x8: 0x220a5822ca787cc5   x9: 0x220a582330a6c645  x10: 0x0000000000179f41  x11: 0x00000000000f4da0
   x12: 0x000000014f600000  x13: 0x0000000104cdfd78  x14: 0x0000000000000000  x15: 0x0000000000000000
   x16: 0x0000000000000148  x17: 0x00000001ffc2c7f0  x18: 0x0000000000000000  x19: 0x0000000000000006
   x20: 0x00000001fadeba80  x21: 0x0000000000000103  x22: 0x00000001fadebb60  x23: 0x000000019fa8d120
   x24: 0x0000000000000000  x25: 0x0000000104298ac8  x26: 0x000060000296c498  x27: 0x0000600002968bc0
   x28: 0x0000000000000000   fp: 0x000000016dd61430   lr: 0x000000019fa84cec
    sp: 0x000000016dd61410   pc: 0x000000019fa4e1b0 cpsr: 0x40001000
   far: 0x0000000104607c28  esr: 0x56000080  Address size fault

Binary Images:
       0x19fa45000 -        0x19fa7dff3 libsystem_kernel.dylib (*) <aebf397e-e2ef-3a49-be58-23d4558511f6> /usr/lib/system/libsystem_kernel.dylib
       0x19fa7e000 -        0x19fa8affb libsystem_pthread.dylib (*) <132084c6-c347-3489-9ac2-fcaad21cdb73> /usr/lib/system/libsystem_pthread.dylib
       0x19f945000 -        0x19f9c5ff3 libsystem_c.dylib (*) <756cd0d2-3241-3a74-8c59-02632dcee221> /usr/lib/system/libsystem_c.dylib
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
       0x19f756000 -        0x19f7e0b63 dyld (*) <487cfdeb-9b07-39bf-bfb9-970b61aea2d1> /usr/lib/dyld

External Modification Summary:
  Calls made by other processes targeting this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by all processes on this machine:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0

VM Region Summary:
ReadOnly portion of Libraries: Total=1.3G resident=0K(0%) swapped_out_or_unallocated=1.3G(100%)
Writable regions: Total=1.5G written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=1.5G(100%)

                                VIRTUAL   REGION 
REGION TYPE                        SIZE    COUNT (non-coalesced) 
===========                     =======  ======= 
Activity Tracing                   256K        1 
Kernel Alloc Once                   32K        1 
MALLOC                           196.2M       21 
MALLOC guard page                   96K        6 
MALLOC_MEDIUM (reserved)         480.0M        4         reserved VM address space (unallocated)
MALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)
STACK GUARD                       56.3M       17 
Stack                             16.5M       17 
VM_ALLOCATE                         80K        2 
VM_ALLOCATE (reserved)           448.0M        1         reserved VM address space (unallocated)
__AUTH                             572K      132 
__AUTH_CONST                      10.4M      277 
__DATA                            14.6M      436 
__DATA_CONST                      20.4M      422 
__DATA_DIRTY                       678K       97 
__FONT_DATA                        2352        1 
__LINKEDIT                       817.9M      146 
__OBJC_CONST                      1161K      107 
__OBJC_RO                         65.4M        1 
__OBJC_RW                         1986K        1 
__TEXT                           477.3M      437 
dyld private memory                256K        1 
mapped file                       29.1M        3 
shared memory                      624K        7 
===========                     =======  ======= 
TOTAL                              3.0G     2139 
TOTAL, minus reserved VM space     1.7G     2139 



-----------
Full Report
-----------

{"app_name":"QGIS","timestamp":"2023-01-11 11:18:46.00 -0500","app_version":"3.28.2","slice_uuid":"00000000-0000-0000-0000-000000000000","build_version":"3.28.2 (exported)","platform":1,"bundleID":"org.qgis.qgis3","share_with_app_devs":0,"is_first_party":0,"bug_type":"309","os_version":"macOS 13.1 (22C65)","roots_installed":0,"name":"QGIS","incident_id":"7B4402E5-5209-4ABD-83AB-5C9B751BE362"}
{
  "uptime" : 1300000,
  "procRole" : "Background",
  "version" : 2,
  "userID" : 501,
  "deployVersion" : 210,
  "modelCode" : "Mac14,2",
  "coalitionID" : 154187,
  "osVersion" : {
    "train" : "macOS 13.1",
    "build" : "22C65",
    "releaseType" : "User"
  },
  "captureTime" : "2023-01-11 11:18:43.4045 -0500",
  "incident" : "7B4402E5-5209-4ABD-83AB-5C9B751BE362",
  "pid" : 45199,
  "translated" : false,
  "cpuType" : "ARM-64",
  "roots_installed" : 0,
  "bug_type" : "309",
  "procLaunch" : "2023-01-11 11:18:42.3638 -0500",
  "procStartAbsTime" : 31809994362957,
  "procExitAbsTime" : 31810019318087,
  "procName" : "QGIS",
  "procPath" : "\/Volumes\/VOLUME\/*\/QGIS.app\/Contents\/MacOS\/QGIS",
  "bundleInfo" : {"CFBundleShortVersionString":"3.28.2","CFBundleVersion":"3.28.2 (exported)","CFBundleIdentifier":"org.qgis.qgis3"},
  "storeInfo" : {"deviceIdentifierForVendor":"A23FFA76-8BE5-5998-8A5C-853D0111BA12","thirdParty":true},
  "parentProc" : "launchd",
  "parentPid" : 1,
  "coalitionName" : "org.qgis.qgis3",
  "crashReporterKey" : "EFA3F67E-B6A3-CBFF-B0CE-8139B7515F3A",
  "throttleTimeout" : 2147483647,
  "wakeTime" : 47222,
  "sleepWakeUUID" : "C5799FA7-6E4C-4286-B156-47BAA862AF96",
  "sip" : "enabled",
  "exception" : {"codes":"0x0000000000000000, 0x0000000000000000","rawCodes":[0,0],"type":"EXC_CRASH","signal":"SIGABRT"},
  "ktriageinfo" : "VM - pmap_enter retried due to resource shortage\nVM - pmap_enter retried due to resource shortage\nVM - pmap_enter retried due to resource shortage\nVM - pmap_enter retried due to resource shortage\nVM - pmap_enter retried due to resource shortage\n",
  "asi" : {"libsystem_c.dylib":["abort() called"]},
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 0,
  "threads" : [{"triggered":true,"id":33188727,"threadState":{"x":[{"value":0},{"value":0},{"value":0},{"value":0},{"value":268437504},{"value":28},{"value":0},{"value":69632},{"value":2452869853505682629},{"value":2452869855219992133},{"value":1548097},{"value":1002912},{"value":5626658816},{"value":4375575928},{"value":0},{"value":0},{"value":328},{"value":8585922544},{"value":0},{"value":6},{"value":8503868032,"symbolLocation":0,"symbol":"_main_thread"},{"value":259},{"value":8503868256,"symbolLocation":224,"symbol":"_main_thread"},{"value":6973608224,"symbolLocation":0,"symbol":"tlv_get_addr"},{"value":0},{"value":4364798664},{"value":105553159701656},{"value":105553159687104},{"value":0}],"flavor":"ARM_THREAD_STATE64","lr":{"value":6973574380},"cpsr":{"value":1073745920},"fp":{"value":6137713712},"sp":{"value":6137713680},"esr":{"value":1442840704,"description":" Address size fault"},"pc":{"value":6973350320,"matchesCrashFrame":1},"far":{"value":4368399400}},"queue":"com.apple.main-thread","frames":[{"imageOffset":37296,"symbol":"__pthread_kill","symbolLocation":8,"imageIndex":0},{"imageOffset":27884,"symbol":"pthread_kill","symbolLocation":288,"imageIndex":1},{"imageOffset":496328,"symbol":"abort","symbolLocation":180,"imageIndex":2},{"imageOffset":4329244688,"symbol":"myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&)","symbolLocation":1052,"imageIndex":3},{"imageOffset":4365879624,"imageIndex":3},{"imageOffset":4365859376,"imageIndex":3},{"imageOffset":4368777316,"symbol":"QMessageLogger::fatal(char const*, ...) const","symbolLocation":56,"imageIndex":3},{"imageOffset":4360433500,"symbol":"QGuiApplicationPrivate::createPlatformIntegration()","symbolLocation":7296,"imageIndex":3},{"imageOffset":4360433536,"symbol":"QGuiApplicationPrivate::createEventDispatcher()","symbolLocation":36,"imageIndex":3},{"imageOffset":4367762772,"symbol":"QCoreApplicationPrivate::init()","symbolLocation":1472,"imageIndex":3},{"imageOffset":4360409836,"symbol":"QGuiApplicationPrivate::init()","symbolLocation":68,"imageIndex":3},{"imageOffset":4339189244,"symbol":"QApplicationPrivate::init()","symbolLocation":52,"imageIndex":3},{"imageOffset":4339189168,"symbol":"QApplication::QApplication(int&, char**, int)","symbolLocation":164,"imageIndex":3},{"imageOffset":4465021124,"symbol":"QgsApplication::QgsApplication(int&, char**, bool, QString const&, QString const&)","symbolLocation":60,"imageIndex":3},{"imageOffset":4329255772,"symbol":"main","symbolLocation":11012,"imageIndex":3},{"imageOffset":24144,"symbol":"start","symbolLocation":2544,"imageIndex":4}]},{"id":33188842,"frames":[{"imageOffset":7704,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":1}]},{"id":33188843,"frames":[{"imageOffset":7704,"symbol":"start_wqthread","symbolLocation":0,"imageIndex":1}]},{"id":33188844,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188845,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188846,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188847,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188848,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188849,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188850,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4672379700,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188851,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188852,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188853,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188854,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188855,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188856,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]},{"id":33188857,"frames":[{"imageOffset":17764,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":0},{"imageOffset":30264,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":1},{"imageOffset":4695137076,"symbol":"blas_thread_server","symbolLocation":360,"imageIndex":3},{"imageOffset":28780,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":1},{"imageOffset":7724,"symbol":"thread_start","symbolLocation":8,"imageIndex":1}]}],
  "usedImages" : [
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6973313024,
    "size" : 233460,
    "uuid" : "aebf397e-e2ef-3a49-be58-23d4558511f6",
    "path" : "\/usr\/lib\/system\/libsystem_kernel.dylib",
    "name" : "libsystem_kernel.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6973546496,
    "size" : 53244,
    "uuid" : "132084c6-c347-3489-9ac2-fcaad21cdb73",
    "path" : "\/usr\/lib\/system\/libsystem_pthread.dylib",
    "name" : "libsystem_pthread.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6972264448,
    "size" : 528372,
    "uuid" : "756cd0d2-3241-3a74-8c59-02632dcee221",
    "path" : "\/usr\/lib\/system\/libsystem_c.dylib",
    "name" : "libsystem_c.dylib"
  },
  {
    "size" : 0,
    "source" : "A",
    "base" : 0,
    "uuid" : "00000000-0000-0000-0000-000000000000"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6970236928,
    "size" : 568164,
    "uuid" : "487cfdeb-9b07-39bf-bfb9-970b61aea2d1",
    "path" : "\/usr\/lib\/dyld",
    "name" : "dyld"
  }
],
  "sharedCache" : {
  "base" : 6969589760,
  "size" : 3434283008,
  "uuid" : "00a1fbb6-43e1-3c11-8483-faf0db659249"
},
  "vmSummary" : "ReadOnly portion of Libraries: Total=1.3G resident=0K(0%) swapped_out_or_unallocated=1.3G(100%)\nWritable regions: Total=1.5G written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=1.5G(100%)\n\n                                VIRTUAL   REGION \nREGION TYPE                        SIZE    COUNT (non-coalesced) \n===========                     =======  ======= \nActivity Tracing                   256K        1 \nKernel Alloc Once                   32K        1 \nMALLOC                           196.2M       21 \nMALLOC guard page                   96K        6 \nMALLOC_MEDIUM (reserved)         480.0M        4         reserved VM address space (unallocated)\nMALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)\nSTACK GUARD                       56.3M       17 \nStack                             16.5M       17 \nVM_ALLOCATE                         80K        2 \nVM_ALLOCATE (reserved)           448.0M        1         reserved VM address space (unallocated)\n__AUTH                             572K      132 \n__AUTH_CONST                      10.4M      277 \n__DATA                            14.6M      436 \n__DATA_CONST                      20.4M      422 \n__DATA_DIRTY                       678K       97 \n__FONT_DATA                        2352        1 \n__LINKEDIT                       817.9M      146 \n__OBJC_CONST                      1161K      107 \n__OBJC_RO                         65.4M        1 \n__OBJC_RW                         1986K        1 \n__TEXT                           477.3M      437 \ndyld private memory                256K        1 \nmapped file                       29.1M        3 \nshared memory                      624K        7 \n===========                     =======  ======= \nTOTAL                              3.0G     2139 \nTOTAL, minus reserved VM space     1.7G     2139 \n",
  "legacyInfo" : {
  "threadTriggered" : {
    "queue" : "com.apple.main-thread"
  }
},
  "trialInfo" : {
  "rollouts" : [
    {
      "rolloutId" : "60da5e84ab0ca017dace9abf",
      "factorPackIds" : {

      },
      "deploymentId" : 240000008
    },
    {
      "rolloutId" : "5fb4245a1bbfe8005e33a1e1",
      "factorPackIds" : {

      },
      "deploymentId" : 240000021
    }
  ],
  "experiments" : [

  ]
}
}

Model: Mac14,2, BootROM 8419.60.44, proc 8:4:4 processors, 24 GB, SMC 
Graphics: Apple M2, Apple M2, Built-In
...
```

Et7f3 commented:
I see source from `3.28.2` and `3.22.14` so maybe a version conflict. Can you test with same version for all sources ? Even better try to avoid duplication.

willcohen commented:
Yeah -- it's two different packages. 3.28.2 is `qgis`, which has a minor release every four months. 3.22 is `qgis-ltr`, which happen every three minor release cycles. The two packages sit in parallel but -- unless I made an oversight -- don't overlap at all. The crash occurs with both `qgis` and `qgis-ltr` though, since they do have similar build processes. My guess is once it's solved for one, that will apply to the other.

imincik commented:
> @imincik any suggestions of how to proceed further based on https://github.com/NixOS/nixpkgs/pull/157862#issuecomment-1379250237 would be much appreciated!

@willcohen , unfortunately I am not a Mac user at all. But I really want to see it running on Mac because of other users. That's why I am very happy that you try to progress.

imincik commented:
Have you tried to build it with debug output ? Something like `-DCMAKE_BUILD_TYPE=Debug` ?

willcohen commented:
.app built with cmake debug crashes with the following, leading me to ask a basic question that's above my knowledge of nix and macos internals -- do these `libsystem` dylibs need to be nix-specific ones? The dylibs being referenced here in the crash reports are definitely in /usr/lib, but I'm not sure how much is supposed to get swapped out from the darwin bootstrap.
```
...
Time Awake Since Boot: 1300000 seconds
Time Since Wake:       65887 seconds

System Integrity Protection: enabled

Notes:
dyld_process_snapshot_create_for_process failed with 5

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Error Formulating Crash Report:
dyld_process_snapshot_create_for_process failed with 5

Kernel Triage:
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage
VM - pmap_enter retried due to resource shortage


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x19fa4e1b0 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x19fa84cec pthread_kill + 288
2   libsystem_c.dylib             	       0x19f9be2c8 abort + 180
3   ???                           	       0x1048d4674 myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&) + 1332
4   ???                           	       0x1072ac948 ???
5   ???                           	       0x1072a7a30 ???
6   ???                           	       0x107570064 QMessageLogger::fatal(char const*, ...) const + 56
7   ???                           	       0x106d7af5c QGuiApplicationPrivate::createPlatformIntegration() + 7296
8   ???                           	       0x106d7af80 QGuiApplicationPrivate::createEventDispatcher() + 36
9   ???                           	       0x107478554 QCoreApplicationPrivate::init() + 1472
10  ???                           	       0x106d752ec QGuiApplicationPrivate::init() + 68
11  ???                           	       0x1052605fc QApplicationPrivate::init() + 52
12  ???                           	       0x1052605b0 QApplication::QApplication(int&, char**, int) + 164
13  ???                           	       0x10ea28228 QgsApplication::QgsApplication(int&, char**, bool, QString const&, QString const&) + 60
14  ???                           	       0x1048d71a4 main + 10976
15  dyld                          	       0x19f75be50 start + 2544

...
Binary Images:
       0x19fa45000 -        0x19fa7dff3 libsystem_kernel.dylib (*) <aebf397e-e2ef-3a49-be58-23d4558511f6> /usr/lib/system/libsystem_kernel.dylib
       0x19fa7e000 -        0x19fa8affb libsystem_pthread.dylib (*) <132084c6-c347-3489-9ac2-fcaad21cdb73> /usr/lib/system/libsystem_pthread.dylib
       0x19f945000 -        0x19f9c5ff3 libsystem_c.dylib (*) <756cd0d2-3241-3a74-8c59-02632dcee221> /usr/lib/system/libsystem_c.dylib
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
       0x19f756000 -        0x19f7e0b63 dyld (*) <487cfdeb-9b07-39bf-bfb9-970b61aea2d1> /usr/lib/dyld
...
```

imincik commented:
> .app built with cmake debug crashes with the following, leading me to ask a basic question that's above my knowledge of nix and macos internals -- do these `libsystem` dylibs need to be nix-specific ones? The dylibs being referenced here in the crash reports are definitely in /usr/lib, but I'm not sure how much is supposed to get swapped out from the darwin bootstrap.
> 
> ```
> ...
> Time Awake Since Boot: 1300000 seconds
> Time Since Wake:       65887 seconds
> 
> System Integrity Protection: enabled
> 
> Notes:
> dyld_process_snapshot_create_for_process failed with 5
> 
> Crashed Thread:        0  Dispatch queue: com.apple.main-thread
> 
> Exception Type:        EXC_CRASH (SIGABRT)
> Exception Codes:       0x0000000000000000, 0x0000000000000000
> 
> Error Formulating Crash Report:
> dyld_process_snapshot_create_for_process failed with 5
> 
> Kernel Triage:
> VM - pmap_enter retried due to resource shortage
> VM - pmap_enter retried due to resource shortage
> VM - pmap_enter retried due to resource shortage
> VM - pmap_enter retried due to resource shortage
> VM - pmap_enter retried due to resource shortage
> 
> 
> Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
> 0   libsystem_kernel.dylib        	       0x19fa4e1b0 __pthread_kill + 8
> 1   libsystem_pthread.dylib       	       0x19fa84cec pthread_kill + 288
> 2   libsystem_c.dylib             	       0x19f9be2c8 abort + 180
> 3   ???                           	       0x1048d4674 myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&) + 1332
> 4   ???                           	       0x1072ac948 ???
> 5   ???                           	       0x1072a7a30 ???
> 6   ???                           	       0x107570064 QMessageLogger::fatal(char const*, ...) const + 56
> 7   ???                           	       0x106d7af5c QGuiApplicationPrivate::createPlatformIntegration() + 7296
> 8   ???                           	       0x106d7af80 QGuiApplicationPrivate::createEventDispatcher() + 36
> 9   ???                           	       0x107478554 QCoreApplicationPrivate::init() + 1472
> 10  ???                           	       0x106d752ec QGuiApplicationPrivate::init() + 68
> 11  ???                           	       0x1052605fc QApplicationPrivate::init() + 52
> 12  ???                           	       0x1052605b0 QApplication::QApplication(int&, char**, int) + 164
> 13  ???                           	       0x10ea28228 QgsApplication::QgsApplication(int&, char**, bool, QString const&, QString const&) + 60
> 14  ???                           	       0x1048d71a4 main + 10976
> 15  dyld                          	       0x19f75be50 start + 2544
> 
> ...
> Binary Images:
>        0x19fa45000 -        0x19fa7dff3 libsystem_kernel.dylib (*) <aebf397e-e2ef-3a49-be58-23d4558511f6> /usr/lib/system/libsystem_kernel.dylib
>        0x19fa7e000 -        0x19fa8affb libsystem_pthread.dylib (*) <132084c6-c347-3489-9ac2-fcaad21cdb73> /usr/lib/system/libsystem_pthread.dylib
>        0x19f945000 -        0x19f9c5ff3 libsystem_c.dylib (*) <756cd0d2-3241-3a74-8c59-02632dcee221> /usr/lib/system/libsystem_c.dylib
>                0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
>        0x19f756000 -        0x19f7e0b63 dyld (*) <487cfdeb-9b07-39bf-bfb9-970b61aea2d1> /usr/lib/dyld
> ...
> ```

My guess will be that it should be nixpkgs provided libs. @NixOS/darwin-maintainers would you please help us with that ?

willcohen commented:
Per https://github.com/NixOS/nixpkgs/commit/70f3863540a8c6350b92b93cb5c13a66e54200d2 it seems like it's okay to refer to the systemwide one -- my read of this is that something QGIS is calling in Qt is leading to this crashing during the invocation of `QGuiApplicationPrivate::createPlatformIntegration()`. @LnL7 or @worldofpeace do either of you have a hunch?

imincik commented:
> Per [70f3863](https://github.com/NixOS/nixpkgs/commit/70f3863540a8c6350b92b93cb5c13a66e54200d2) it seems like it's okay to refer to the systemwide one -- my read of this is that something QGIS is calling in Qt is leading to this crashing during the invocation of `QGuiApplicationPrivate::createPlatformIntegration()`. @LnL7 or @worldofpeace do either of you have a hunch?

@willcohen , quite old but maybe - https://bugreports.qt.io/browse/QTBUG-48015

viraptor commented:
darwin is not completely pure unfortunately. The `/usr/lib/libSystem` and probably some other references will be present in nix builds.

Et7f3 commented on pkgs/applications/gis/qgis/unwrapped.nix:
```diff
@@ -132,31 +138,76 @@ in mkDerivation rec {
     "-DWITH_3D=True"
-    "-DWITH_PDAL=TRUE"
   ] ++ lib.optional (!withWebKit) "-DWITH_QTWEBKIT=OFF"
+    ++ lib.optional stdenv.isLinux "-DWITH_PDAL=TRUE"
```
Why this change ? it seems to build fine on Mac with this.

willcohen replied:
Good call -- it was broken a long time ago and I just assumed that was still true. Fixed since October with https://github.com/NixOS/nixpkgs/pull/195216. Will revise.

Et7f3 commented on pkgs/applications/gis/qgis/unwrapped.nix:
```diff
@@ -132,31 +138,76 @@ in mkDerivation rec {
+             $out/Applications/QGIS.app/Contents/Frameworks/qgisgrass8.framework/qgisgrass8
+      do
+        echo "Manually running install_name_tool on $f"
+        install_name_tool -change "@loader_path/../lib/libqscintilla2_qt5.dylib" "${qscintilla}/lib/libqscintilla2_qt5.dylib" $f
```
how did you get this list ? can it be parsed and looped over ? it should be more resistant to future upgrade

willcohen replied:
This was, unfortunately, done by me repeatedly building and running the .app, having it crash due to a linking error, and repeating over and over. I did at least simplify it so it can loop over the dylibs I know about. It seems like the [original build script does it with a little less targeting](https://github.com/qgis/QGIS-Mac-Packager/blob/master/qgis_bundle/qgis_bundle.bash), but it's true that this is moderately fragile. I'm open to suggestions for how to improve it -- fixDarwinDylib names was too blunt an instrument and tried to operate on lots of files in $out, and wasn't able to deal with the Frameworks, so I ended up just doing this manually.

Et7f3 commented on pkgs/applications/gis/qgis/unwrapped.nix:
```diff
@@ -132,31 +138,76 @@ in mkDerivation rec {
+      do
+        echo "Manually running install_name_tool on $f"
+        install_name_tool -change "@loader_path/../lib/libqscintilla2_qt5.dylib" "${qscintilla}/lib/libqscintilla2_qt5.dylib" $f
+        install_name_tool -change "@loader_path/../lib/libqt5keychain.dylib" "${qtkeychain}/lib/libqt5keychain.dylib" $f
```
```suggestion
        install_name_tool -change "@loader_path/../lib/libqt5keychain.dylib" "${qtkeychain}/lib/libqt5keychain.dylib" "$f"
```

same for all other

willcohen replied:
Good catch -- will revise.

Et7f3 reviewed (commented):
I tried to debug `nix develop -i .#qgis.passthru.unwrapped` but I get error with missing nose2 import. Can you see your log to see if the tests fail on your machine ? (I have Darwin sandbox mode)

willcohen commented:
> nix develop -i .#qgis.passthru.unwrapped

I'm not sure that this will ever work with sandbox enabled. Before even anything about tests failing, with sandbox enabled I get `ld: warning: directory not found for option '-F/Library/Frameworks'` failures in the build. My post-fixup changes end up relinking all the dylibs around so that they're linked based on the symbolic links that get generated in the end, but it seems like keeping the build that pure might be a version 2.0 kind of task for this derivation.

Et7f3 commented:
I think that without sandboxing you might get a dependencies that mismatch and it can cause failure. Once I got a build issue because of my locale that wasn't unset. But ok I will un sandbox for this derivation. I was able to launch some test in sandbox mode.

willcohen committed:
qgis, qgis-ltr: fix dependencies and linking on darwin

willcohen commented:
Dusted this off and I think I got it working!
`qgis` and `qgis-ltr` now working with today's changes. 

I'd love another mac user to test it out to confirm it's not just me.

<img width="1368" height="943" alt="Screenshot 2026-01-14 at 7 58 38 PM" src="https://github.com/user-attachments/assets/9310dc52-e279-4d03-8849-ba31f9bf0448" />

willcohen commented:
Lovely. Working on it but might take a few rounds -- the darwin packaging changes here need some tweaks for Qt6 (which is to be expected). If anyone can confirm that 3.x is working okay here we could still merge this independent of what ends up upstream -- which can then hopefully easily be adapted as we migrate to 4 here in the near future.

imincik commented:
> If anyone can confirm that 3.x is working okay

Unfortunately, I don't have any Darwin machine to test but I'll promote this PR in QGIS community.

Xpirix commented:
Thank you so much @willcohen . I have tested it on an M1 Pro, and it seems to be working nicely. Here are some random, simple tests that I ran:

- [x] Install a plugin from the repo
- [x] Open a project
- [x] Make some edits
- [x] Open attribute table
- [x] Render a Map

Cc @imincik 

<img width="1999" height="1162" alt="image" src="https://github.com/user-attachments/assets/224e2966-eafe-4611-9226-55c06620e4f0" />

imincik commented on pkgs/applications/gis/qgis/unwrapped.nix:
```diff
@@ -183,18 +196,102 @@ mkDerivation rec {
+          --prefix PATH : ${lib.makeBinPath [ grass ]}
+      done
+    ''
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
```
Can we extract `fix_binary` function to separate file and reuse it for both latest and ltr package ?

imincik commented on pkgs/applications/gis/qgis/unwrapped-ltr.nix:
```diff
@@ -183,18 +200,102 @@ mkDerivation rec {
+            mv $out/QGIS.app $out/Applications/
+            ln -s $out/Applications/QGIS.app/Contents/MacOS/QGIS $out/bin/qgis
+
+            SHORT_VERSION=$(echo "${version}" | cut -d. -f1,2)
```
```suggestion
            SHORT_VERSION="${lib.versions.majorMinor version }"
```

imincik reviewed (changes requested):
I requested few changes but thanks for great work !

siraben commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 157862`
Commit: `d5a42c0770963e990ef01d1d731d8d54e86e4200`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>qgis</li>
    <li>qgis-ltr</li>
  </ul>
</details>
